casper/network.py

```diff
@@ -10,74 +11,62 @@ def __init__(self, validator_set, protocol=BlockchainProtocol, force_justify_mes
             self._collect_initial_messages(),
             protocol.initial_message(None)
         )
+        self.message_queues = {
+            validator: PriorityQueue()
+            for validator in self.validator_set
+        }
+        self.time = 0
 
         self.force_justify_messages = force_justify_messages
 
-    def propagate_message_to_validator(self, message, validator):
-        """Propagate a message to a validator."""
-        assert message.hash in self.global_view.justified_messages, (
-            "...expected only to propagate messages "
-            "from the global view")
-        assert validator in self.validator_set, "...expected a known validator"
-
-        validator.receive_messages(set([message]))
-
-        # HACK TO SEND ALL MESSAGES BEFORE NETWORK REWORK
-        if self.force_justify_messages:
-            self._propagate_messages_needed_to_justify(message, validator)
-
-    def get_message_from_validator(self, validator):
-        """Get a message from a validator."""
-        assert validator in self.validator_set, "...expected a known validator"
-
-        new_message = validator.make_new_message()
-        self.global_view.add_messages(set([new_message]))
-        assert new_message.hash in self.global_view.justified_messages
-
-        return new_message
-
-    def view_initialization(self, view):
-        """Initalizes all validators with all messages in some view."""
-        messages = view.justified_messages.values()
-
-        self.global_view.add_messages(messages)
-
+    def delay(self, sender, receiver):
+        '''Must be defined in child class.
+        Returns delay of next message for sender to receiver'''
+        raise NotImplementedError
+
+    #
+    # Validator API to Network
```

Comment:
API written in such a way that Validators could easily be put into their own process. The SimRunner is essentially a single-threaded simulation runner. Could now write a multi-threaded sim-runner where network is it's own thread and each validator it's own thread. Validators would just interface with this API.

The `send` method is interesting. If a validator can call `send` then that implies that they are directly sending messages to peers. `send` is also cool because a validator can send anyone's message. If we had validators with limited peers, we could model a gossip protocol where validators attempt to send any message they get to their known peers.